documentation/smg/04_building_with_cmake.md

```diff
@@ -1,38 +1,151 @@
-You should pick a compiler that is compatible with your Matlab version, e.g.
-Matlab R2017b is not compatible with Visual Studio 2019, so choose Visual Studio 2017.
-4. Make sure to also select your platform/architecture, this should also match
-your Matlab version: if you installed a 64-bit version of Matlab choose x64 as
-the platform.
-5. Now click finish in the dialogue. CMake will find the Matlab compiler and
-various libraries.
-6. Now click generate. The build files should be generated inside the DLL
-directory. On Windows you can open the `<Horace Root>/build/Horace.sln` file in
-Visual Studio and build the targets. On linux you can `cd` into the
-`<Horace Root>/build/` directory and run `make` depending on the generator you
-selected.
-
-To do steps 3-6 using the command line you can use the following commands (for
+Create this directory if it does not exist.
+
+3. Click configure.
+When the dialogue appears select your desired generator.
+You should pick a compiler that is compatible with your Matlab version,
+see the [version compatibility matrix](./09_version_compatibility.md)
+(whilst the version compatibility matrix is a good guide,
+the mex libraries will likely work if built using newer compiler versions).
+
+4. Make sure to also select your platform/architecture,
+this should also match your Matlab version:
+if you installed a 64-bit version of Matlab choose x64 as the platform.
+
+5. Now click finish in the dialogue.
+CMake will find the Matlab compiler and various libraries.
+
+6. Now click generate.
+The build files should be generated inside the DLL directory.
+On Windows you can open the `<Horace Root>/build/Horace.sln` file in Visual
+Studio and build the targets.
+
+7. The built mex files will be written into `horace_core/DLL`.
+
+To do steps 2-6 using the command line you can use the following commands (on
 linux use `-G "Unix Makefiles"`):
 
 `$ cmake -S <Horace Root> -B <Horace Root>/build -G "Visual Studio 15 2017 Win64"`
 
-`$ cmake --build <Horace Root>/build`
+`$ cmake --build <Horace Root>/build --config Release`
+
+There are known issues running mex files compiled using old versions of CMake
+on new Matlab versions.
+If you get an error like `Invalid MEX-file: Gateway function is missing`,
+try updating CMake.
+CMake v3.14 has been confirmed to work for Matlab R2019b.
+
+### Troubleshooting
+
+#### Invalid MEX-file
+
+If, when running mex files built using one of the above processes,
+you get an error like `Invalid MEX-file: Gateway function is missing`,
+this means the mex library is not exporting the correct symbols.
+On Windows the mex library (a DLL) should export `mexFunction` and only that.
+On Linux, Matlab provides a `.map` file that dictates what symbols to export
+from the shared library.
+If Matlab was installed in the standard directory on a 64-bit system,
+the path to this file will be
+`/usr/local/MATLAB/R2018b/extern/lib/glnxa64/mexFunction.map`.
+
```

Comment:
Run the paragraph starting "If Matlab" on from the previous paragraph without a '\n' - it is part of the liinux instruction, not a new paragraph which might refer to Windows or Linux.

Reply:
Realising that the '\n' is markdown only. But separating Windows and Linux with "\n\n" could be a good idea.